### PR TITLE
24143 new health check

### PR DESCRIPTION
**Pull request by vipin14904**
24143 - ENS - Health Check

### CODE REVIEW

**Comment by vipin14904** [comment]
Conflict Resolved, You can start review.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Logging/EventIds.cs [hunk 1]
```diff
@@ -35,7 +36,19 @@ public enum EventIds
...
+        EventHubLoggingIsHealthy = 900009,
+        /// <summary>
+        /// 900010 - Event hub for exchange set service is unhealthy.
+        /// </summary>
+        EventHubLoggingIsUnhealthy = 900010,
+        /// <summary>
+        /// 900011 -  Event data for exchange set service event hub health check.
+        /// </summary>
+        EventHubLoggingEventDataForHealthCheck = 900011
     }
```
Comment (usha13529) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Logging/EventIds.cs:52]: Change comments as below:
/// 900009 - Event hub for external notification service is healthy.
/// 900010 - Event hub for external notification service is unhealthy.
/// 900011 - Event data for external notification service event hub health check.

**Review by usha13529 — CHANGES_REQUESTED** [review]
Please fix comments text.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Logging/EventIds.cs (hunk 1 above)
Comment (vipin14904) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Logging/EventIds.cs:52]: Changes are done as requested.

**Review by usha13529 — APPROVED** [review]
LGTM.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+        public EventHubLoggingHealthClient(IOptions<EventHubLoggingConfiguration> eventHubLoggingConfiguration)
+        {
+            this._eventHubLoggingConfiguration = eventHubLoggingConfiguration;
+        }
+
+        public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        {
+            var eventHubProducerClient = new EventHubProducerClient(_eventHubLoggingConfiguration.Value.ConnectionString, _eventHubLoggingConfiguration.Value.EntityPath);
+            using EventDataBatch eventBatch = await eventHubProducerClient.CreateBatchAsync();
```
Comment (barrie-cooper) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs:29]: `CreateBatchAsync` has an overload with cancellation support so please use it:
```suggestion
            using EventDataBatch eventBatch = await eventHubProducerClient.CreateBatchAsync(cancellationToken);
```

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            eventBatch.TryAdd(new EventData(Encoding.UTF8.GetBytes(EventIds.EventHubLoggingEventDataForHealthCheck.ToEventId() + " of Event Hub")));
+            try
+            {         
+                await eventHubProducerClient.SendAsync(eventBatch);
```
Comment (barrie-cooper) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs:33]: `SendAsync` also has an overload with cancellation support:
```suggestion
                await eventHubProducerClient.SendAsync(eventBatch, cancellationToken);
```

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                return HealthCheckResult.Healthy("Event hub is healthy");
+            }
+            catch (Exception ex)
+            {
+                return HealthCheckResult.Unhealthy("Event hub is unhealthy", new Exception(ex.Message));
+            }
+            finally
+            {
+                await eventHubProducerClient.CloseAsync();
```
Comment (barrie-cooper) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs:42]: `CloseAsync` also has an overload with cancellation support:
```suggestion
                await eventHubProducerClient.CloseAsync(cancellationToken);
```

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+namespace UKHO.ExternalNotificationService.Common.HealthCheck
+{
+    [ExcludeFromCodeCoverage]
+    public class EventHubLoggingHealthClient : IEventHubLoggingHealthClient
+    {
+        private readonly IOptions<EventHubLoggingConfiguration> _eventHubLoggingConfiguration;
+
+        public EventHubLoggingHealthClient(IOptions<EventHubLoggingConfiguration> eventHubLoggingConfiguration)
+        {
+            this._eventHubLoggingConfiguration = eventHubLoggingConfiguration;
```
Comment (barrie-cooper) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs:23]: `this.` is not needed, please remove:
```suggestion
            _eventHubLoggingConfiguration = eventHubLoggingConfiguration;
```

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/IEventHubLoggingHealthClient.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+﻿
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+
+namespace UKHO.ExternalNotificationService.Common.HealthCheck
+{
+    public interface IEventHubLoggingHealthClient
+    {
+        public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default);
```
Comment (barrie-cooper) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/IEventHubLoggingHealthClient.cs:10]: Access specifier not needed in an interface, please remove:
```suggestion
        Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default);
```

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthCheck.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+        public EventHubLoggingHealthCheck(IEventHubLoggingHealthClient eventHubLoggingHealthClient, ILogger<EventHubLoggingHealthCheck> logger)
+        {
+            this._eventHubLoggingHealthClient = eventHubLoggingHealthClient;
+            this._logger = logger;
+        }
+
+        public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        {
+            var healthCheckResult = await _eventHubLoggingHealthClient.CheckHealthAsync(context);
```
Comment (barrie-cooper) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthCheck.cs:22]: Please pass `cancellationToken`:
```suggestion
            var healthCheckResult = await _eventHubLoggingHealthClient.CheckHealthAsync(context, cancellationToken);
```

**Review by barrie-cooper — COMMENTED** [review]
A few comments to address please

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthCheck.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+{
+    public class EventHubLoggingHealthCheck : IHealthCheck
+    {
+        private readonly IEventHubLoggingHealthClient _eventHubLoggingHealthClient;
+        private readonly ILogger<EventHubLoggingHealthCheck> _logger;
+
+        public EventHubLoggingHealthCheck(IEventHubLoggingHealthClient eventHubLoggingHealthClient, ILogger<EventHubLoggingHealthCheck> logger)
+        {
+            this._eventHubLoggingHealthClient = eventHubLoggingHealthClient;
+            this._logger = logger;
```
[the remark this answers is not in the document]
Comment (vipin14904) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthCheck.cs:17]: Yes. Changes Done as requested

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs (hunk 2 above)
Comment (vipin14904) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs:29]: Yes, I have passed cancellationToken as parameter

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs (hunk 3 above)
Comment (vipin14904) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs:33]: Yes, I have passed cancellationToken as parameter

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs (hunk 4 above)
Comment (vipin14904) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs:42]: Yes, I have passed cancellationToken as parameter

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs (hunk 5 above)
Comment (vipin14904) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs:23]: Yes. Changes Done as requested

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/IEventHubLoggingHealthClient.cs (hunk 6 above)
Comment (vipin14904) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/IEventHubLoggingHealthClient.cs:10]: Access specifier removed

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthCheck.cs (hunk 7 above)
Comment (vipin14904) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthCheck.cs:22]: Yes, I have passed cancellationToken as parameter

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using UKHO.ExternalNotificationService.Common.Configuration;
+using UKHO.ExternalNotificationService.Common.Logging;
+
+namespace UKHO.ExternalNotificationService.Common.HealthCheck
+{
+    [ExcludeFromCodeCoverage]
```
[the remark this answers is not in the document]
Comment (vipin14904) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs:16]: We need actual event hub credentials to write a unit test case for this class method, That's why we were excluded from code coverage.

[the remark this answers is not in the document]
Comment (barrie-cooper) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/HealthCheck/EventHubLoggingHealthClient.cs:16]: Thanks, I guessed there was a good reason, I just couldn't think of it.

**Review by barrie-cooper — APPROVED** [review]
LGTM, thanks for changes.

**Pull request merged**